zk_toolbox/crates/zk_inception/src/commands/portal.rs

```diff
@@ -0,0 +1,240 @@
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct L1NetworkConfig {
+    pub id: u64,
+    pub name: String,
+    pub network: String,
+    #[serde(rename = "nativeCurrency")]
+    pub native_currency: NativeCurrency,
+    #[serde(rename = "rpcUrls")]
+    pub rpc_urls: RpcUrls,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct NativeCurrency {
+    pub name: String,
+    pub symbol: String,
+    pub decimals: u8,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct RpcUrls {
+    pub default: RpcUrlConfig,
+    pub public: RpcUrlConfig,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct RpcUrlConfig {
+    pub http: Vec<String>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct TokenConfig {
+    pub address: String,
+    pub symbol: String,
+    pub decimals: u8,
+    #[serde(rename = "l1Address", skip_serializing_if = "Option::is_none")]
+    pub l1_address: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub name: Option<String>,
+}
+
+pub fn create_hyperchain_config(
+    ecosystem_config: &EcosystemConfig,
+    chain_config: &ChainConfig
+) -> anyhow::Result<HyperchainConfig> {
+    // Get L2 RPC URL from general config
+    let general_config = chain_config.get_general_config()?;
+    let rpc_url = &general_config
+        .api_config
+        .as_ref()
+        .context("api_config")?
+        .web3_json_rpc
+        .http_url;
+    // Check if L1 network is public
+    let public_l1_network_id = match chain_config.l1_network {
+        L1Network::Sepolia | L1Network::Mainnet => Some(chain_config.l1_network.chain_id()),
+        _ => None,
+    };
+    let l1_network = if public_l1_network_id.is_none() {
+        // Initialize non-public L1 network
+        let secrets_config = chain_config.get_secrets_config()?;
+        let l1_rpc_url = secrets_config.l1.as_ref().context("l1")?.l1_rpc_url.expose_str();
+        Some(L1NetworkConfig {
+            id: chain_config.l1_network.chain_id(),
+            name: chain_config.l1_network.to_string(),
+            network: chain_config.l1_network.to_string().to_lowercase(),
+            native_currency: NativeCurrency {
+                name: "Ether".to_string(),
+                symbol: "ETH".to_string(),
+                decimals: 18,
+            },
+            rpc_urls: RpcUrls {
+                default: RpcUrlConfig {
+                    http: vec![l1_rpc_url.to_string()],
+                },
+                public: RpcUrlConfig {
+                    http: vec![l1_rpc_url.to_string()],
+                },
+            },
+        })
+    } else {
+        None
+    };
+    // TODO: use bridge contract to derive base token
+    // TODO: add ERC20 tokens
+    // Tokens: add ETH token
+    let tokens = vec![
+        TokenConfig {
+            address: L2_BASE_TOKEN_ADDRESS.to_string(),
+            l1_address: Some("0x0000000000000000000000000000000000000000".to_string()),
```

Comment:
it's not true for chains with custom base token.

Reply:
now getting it from config
interesting edge-case - what happens when L1 base token has e.g. 6 decimals, I assume L2 base token would still have 18? (e.g. some [contracts](https://github.com/matter-labs/era-contracts/blob/8670004d6daa7e8c299087d62f1451a3dec4f899/l2-contracts/contracts/bridge/L2WrappedBaseToken.sol#L43) rely on?)